File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummyHead=new ListNode(0);
        dummyHead->next=head;
        ListNode* fast=dummyHead;
        ListNode* slow=dummyHead;
        while(n--&&fast->next!=NULL){
            fast=fast->next;    
        }
        fast=fast->next;    //fast 再多走一步，slow 最後會走到要刪除的前一個節點
        while(fast!=NULL){
            fast=fast->next;
            slow=slow->next;
        }
        slow->next=slow->next->next;
        return dummyHead->next;

    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummyHead=ListNode(0)
        dummyHead.next=head
        fast=dummyHead
        slow=dummyHead
        while n>0 and fast is not None:
            fast=fast.next
            n-=1
        fast=fast.next  #fast 多走一步，slow 就會指到要刪的前一個節點
        while fast is not None:
            fast=fast.next
            slow=slow.next
        slow.next=slow.next.next
        return dummyHead.next